wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1064 @@
+        const unsigned char* priv, unsigned int privSz,
+        const unsigned char* otherPubKeyDer, unsigned int otherPubKeySz,
+        unsigned char* out, word32* outlen,
+        void* ctx)
+{
+
+    WOLFSSL_MSG("IOTSAFE: Called wolfIoT_ecc_dh_agree\n");
+    (void)ctx;
+    (void)ssl;
+    (void)key;
+    (void)priv;
+    (void)privSz;
+    (void)otherPubKeyDer;
+    (void)otherPubKeySz;
+    (void)out;
+    (void)outlen;
+
+    /* TODO */
+    return 0;
+}
+#endif /* ! NO_DH */
+
+#endif /* HAVE_PK_CALLBACKS */
+
+/* Public API for ecc */
+
+int ecc_import_public_iotsafe(ecc_key *key, byte key_id)
+{
+    return iotsafe_get_public_key(key_id, key);
+}
+
+int ecc_export_public_iotsafe(ecc_key *key, byte key_id)
+{
+    return iotsafe_put_public_key(key_id, key);
+}
+
+int ecc_sign_hash_iotsafe(byte *in, word32 inlen, byte *out,
+        word32 *outlen, byte key_id)
+{
+
+    return iotsafe_sign_hash(key_id, IOTSAFE_HASH_SHA256, IOTSAFE_SIGN_ECDSA,
+            in, inlen, out, outlen);
+}
+int ecc_verify_hash_iotsafe(byte *sig, word32 siglen, byte *hash,
+        word32 hashlen, int *res, byte key_id)
+{
+    word32 r_size = IOTSAFE_ECC_KSIZE;
+    word32 s_size = IOTSAFE_ECC_KSIZE;
+    *res = 0;
+    byte sig_raw[IOTSAFE_ECC_KSIZE * 2];
+    /* Convert ECC signature into R,S */
+    wc_ecc_sig_to_rs(sig, siglen, sig_raw, &r_size, sig_raw
+            + IOTSAFE_ECC_KSIZE, &s_size);
+    return iotsafe_verify_hash(key_id, IOTSAFE_HASH_SHA256, IOTSAFE_SIGN_ECDSA,
+            hash, hashlen, sig_raw, 2 * IOTSAFE_ECC_KSIZE, res);
+}
+int ecc_gen_k_iotsafe(byte key_id)
+{
+    return iotsafe_gen_keypair(key_id);
+}
+
+#endif /* HAVE_ECC */
+
+#ifndef NO_RSA
+
+
+#endif /* !NO_RSA */
+/* API for GetRandom, used as custom Rand seed or block.
+ * Returns 0 on success. */
+int wolfIoTSafe_GetRandom(unsigned char *out, word32 sz)
+{
+    return iotsafe_getrandom(out, sz);
+}
+
+/* API for GetCert (proxy for Read File / Close File)
+ */
+int wolfIoTSafe_GetCert(uint8_t id, unsigned char *output, unsigned long sz)
+{
+    memset(output, 0, sz);
+    return iotsafe_readfile(id, output, (word32)sz);
+}
+
+/* API to set target specific I/O callbacks */
+void wolfIoTSafe_SetCSIM_read_cb(wolfSSL_IOTSafe_CSIM_read_cb rf)
+{
+    csim_read_cb = rf;
+}
+
+void wolfIoTSafe_SetCSIM_write_cb(wolfSSL_IOTSafe_CSIM_write_cb wf)
+{
+    csim_write_cb = wf;
+}
+
+/* API to equip target wolfSSL CTX to the IoT-Safe subsystem. */
+void wolfSSL_CTX_iotsafe_enable(WOLFSSL_CTX *ctx)
+{
+
+    if ( !wolfIoT_initialized)
+    {
+        iotsafe_init();
+    }
+    #if defined(HAVE_PK_CALLBACKS)
+        #ifdef HAVE_ECC
+        wolfSSL_CTX_SetEccSignCb(ctx, wolfIoT_ecc_sign);
+        wolfSSL_CTX_SetEccVerifyCb(ctx, wolfIoT_ecc_verify);
+        wolfSSL_CTX_SetEccKeyGenCb(ctx, wolfIoT_ecc_keygen);
+        wolfSSL_CTX_SetEccSharedSecretCb(ctx, wolfIoT_ecc_shared_secret);
+        wolfSSL_CTX_SetDhAgreeCb(ctx, wolfIoT_dh_agree);
```

Comment:
You might consider not setting the DH agree callback until it is supported. Also it is for NO_DH not HAVE_ECC.

Reply:
OK for the guards, but I'd rather keep the unimplemented DH stub for now to avoid 'unused' warings, and make clear that the callback is offloaded to the module after wolfSSL_iotsafe_on(). I changed the return value to `WC_HW_E` so it's easy to spot it's not implemented yet.